src/js_sync.hpp

```diff
@@ -166,6 +230,102 @@ void UserClass<T>::logout(ContextType ctx, FunctionType, ObjectType this_object,
 }
 
 template<typename T>
+void UserClass<T>::set_permission(ContextType ctx, FunctionType, ObjectType this_object, size_t argc, const ValueType args[], ReturnValue &) {
+    validate_argument_count(argc, 4);
+    Permission::AccessLevel access;
+    std::string url = Value::validated_to_string(ctx, args[0], "url");
+    std::string str_access = Value::validated_to_string(ctx, args[1], "access");
+    std::string user_id = Value::validated_to_string(ctx, args[2], "userId");
+    FunctionType callback = Value::validated_to_function(ctx, args[3], "callback");
+
+    if (str_access == "Read") access = Permission::AccessLevel::Read;
+    else if (str_access == "Write") access = Permission::AccessLevel::Write;
+    else if (str_access == "Admin") access = Permission::AccessLevel::Admin;
+    else throw std::runtime_error("Invalid access level");
+
+    Protected<GlobalContextType> protected_ctx(Context<T>::get_global_context(ctx));
+    Protected<FunctionType> protected_callback(ctx, callback);
+    Permissions::set_permission(*get_internal<T, UserClass<T>>(this_object),
+        { url, access, user_id },
+        [protected_ctx, protected_callback](auto ex) {
+            HANDLESCOPE
+            ValueType error;
+            try {
+                if (ex) std::rethrow_exception(ex);
+                error = Value::from_undefined(protected_ctx);
+            } catch(const std::exception& e) {
+                error = Exception<T>::value(protected_ctx, e.what());
+            }
+            Function::call(protected_ctx, protected_callback, 1, &error);
+        },
+        [ctx] (auto user, auto url) {
+            Realm::Config config;
+            populate_sync_config_impl<T>(ctx, config, std::move(user), url, [](auto, auto) {});
+            return config;
```

Comment:
I can't quite follow the flow here (but it looks like it's correct, judging from the tests) -- why are we giving `set_permission` a callback that creates a configuration object?